### PR TITLE
interp: use functions for exit status checks

### LINKED ISSUE #426: interp: replace ExitStatus with functions
With extension of errors package in go 1.13 comparison with error type becomes even more clunky than it was before. Now user can wrap errors with `fmt.Errorf("my shell script failed: %w", err)` user will have to change their checks from this
```
code, ok := err.(interp.ExitStatus)
if !ok && err != nil {
   t.Fatal(err)
}

if code != 1 {
    t.Fatalf("expected exit code 1, got: %d", code)
}
```
to this
```
var code *interp.ExitStatus
if err == nil || !errors.As(err, &code) {
    t.Fatalf("expected ExitStatus error, got %v", err)
}

if code != 1 {
    t.Fatalf("expected exit code 1, got: %d", code)
}
```

I suggest making 2 funcs instead `func NewExitStatus(code int) error` `func IsExitStatus(err error) (code int, ok bool)` and unexport error type.
with this check changes to 
```
code, ok := interp.IsExitStatus(err)
if !ok && err != nil {
   t.Fatal(err)
}

if code != 1 {
    t.Fatalf("expected exit code 1, got: %d", code)
}
```

While all these examples look similar, I find first and third much easier to understand and I spent less time writing them, than second one.

There is also solves potential issue with forward compatibility, say you want to support plan9 and need to change error from `type ExitStatus uint8` to `type ExitStatus struct {Code uint8, Str string}` with types that would be a breaking change as users will need to update all their error checks. With functions you simply update 2 functions to recognize new error type add 2 more functions for new type, thats not a breaking change.

### PR DESCRIPTION
**Pull request by kaey**
Unexport ExitStatus.
Replace with NewExitStatus() and IsExitStatus().

Fixes #426.

### CODE REVIEW

File: interp/handler.go [hunk 1]
```diff
@@ -53,9 +53,8 @@ type HandlerContext struct {
 // called for all CallExpr nodes where the first argument is neither a
 // declared function nor a builtin.
 //
-// Use a return error of type ExitStatus to set the exit status. A nil error has
-// the same effect as ExitStatus(0). If the error is of any other type, the
-// interpreter will come to a stop.
+// Returning nil error sets commands exit status to 0. Other exit statuses
+// can be set with NewExitStatus(). Any other error will halt an interpreter.
```
Comment (mvdan) [inline interp/handler.go]: remove the ()

File: interp/interp.go [hunk 2]
```diff
@@ -595,10 +596,25 @@ func (r *Runner) handlerCtx(ctx context.Context) context.Context {
...
+	return exitStatus(status)
+}
+
+// IsExitStatus checks whether error contains exit status and returns it.
+func IsExitStatus(err error) (status uint8, ok bool) {
+	var s exitStatus
+	if errors.As(err, &s) {
+		return uint8(s), true
+	}
+
```
Comment (mvdan) [inline interp/interp.go]: remove this empty line

File: interp/interp.go [hunk 3]
```diff
@@ -595,10 +596,25 @@ func (r *Runner) handlerCtx(ctx context.Context) context.Context {
...
 
-func (s ExitStatus) Error() string { return fmt.Sprintf("exit status %d", s) }
+func (s exitStatus) Error() string { return fmt.Sprintf("exit status %d", s) }
+
+// NewExitStatus creates an error which contains specified exit status code.
+func NewExitStatus(status uint8) error {
+	return exitStatus(status)
+}
+
+// IsExitStatus checks whether error contains exit status and returns it.
```
Comment (mvdan) [inline interp/interp.go]: contains an exit status

File: interp/interp.go [hunk 4]
```diff
@@ -607,7 +623,8 @@ func (r *Runner) setErr(err error) {
 }
 
 // Run interprets a node, which can be a *File, *Stmt, or Command. If a non-nil
-// error is returned, it will typically be of type ExitStatus.
+// error is returned, it will typically contain commands exit status,
+// which can be retrieved with IsExitStatus().
```
Comment (mvdan) [inline interp/interp.go]: remove the ()

**Comment by kaey** [comment]
go 1.12 tests fail. I can switch to x/errors if you want to support it?

File: interp/interp.go [hunk 5]
```diff
@@ -595,10 +596,25 @@ func (r *Runner) handlerCtx(ctx context.Context) context.Context {
...
-// ExitStatus is a non-zero status code resulting from running a shell node.
-type ExitStatus uint8
+// exitStatus is a non-zero status code resulting from running a shell node.
+type exitStatus uint8
 
-func (s ExitStatus) Error() string { return fmt.Sprintf("exit status %d", s) }
+func (s exitStatus) Error() string { return fmt.Sprintf("exit status %d", s) }
+
+// NewExitStatus creates an error which contains specified exit status code.
+func NewExitStatus(status uint8) error {
```
[the remark this answers is not in the document]
Comment (kaey) [reply interp/interp.go]: done

**Comment by mvdan** [comment]
Yes, we don't want to be breaking Go 1.12 before 1.14 is out. Adding x/errors temporarily is okay.

**Comment by kaey** [comment]
Switched to x/errors

**Pull request merged**